src/ApiClient.php

```diff
@@ -0,0 +1,181 @@
+     * @param RequestFactoryInterface $requestFactory
+     * @param StreamFactoryInterface  $streamFactory
+     * @param UriFactoryInterface     $uriFactory
+     * @param array                   $configuration
+     */
+    public function __construct(
+        ClientInterface $httpClient,
+        RequestFactoryInterface $requestFactory,
+        StreamFactoryInterface $streamFactory,
+        UriFactoryInterface $uriFactory,
+        array $configuration
+    ) {
+        $this->uriFactory = $uriFactory;
+        $this->configuration = $configuration;
+        $this->createContainer(
+            $httpClient,
+            $requestFactory,
+            $streamFactory,
+            $uriFactory,
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function requestToken(): Token
+    {
+        /** @var TokenEndpoint $endpoint */
+        $endpoint = $this->container->get('auth');
+
+        return $endpoint->execute();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function checkConnection(): bool
+    {
+        /** @var MainEndpoint $endpoint */
+        $endpoint = $this->container->get('main');
+        $endpoint->setToken($this->getToken());
+
+        return $endpoint->execute();
+    }
+
+    /**
+     * @param ClientInterface     $httpClient
+     * @param RequestFactoryInterface $requestFactory
+     * @param StreamFactoryInterface  $streamFactory
+     * @param UriFactoryInterface     $uriFactory
+     */
+    private function createContainer(
+        ClientInterface $httpClient,
+        RequestFactoryInterface $requestFactory,
+        StreamFactoryInterface $streamFactory,
+        UriFactoryInterface $uriFactory
+    ): void {
+        $container = new Container();
+
+        $container->add('token_config', new LiteralArgument($this->extractConfigValues([
+            'username',
+            'password',
+            'client',
+        ])));
+
+        // All services are not shared, meaning that a new instance is
+        // created every time the service is requested from the container.
+        // We're doing this because such a service might be called more than
+        // once during the lifetime of a request, so internals set in a previous
+        // usage may leak into the later usages.
+        $container->add('multipartStreamBuilder', MultipartStreamBuilder::class)
+            ->addArgument($streamFactory);
+        $container->add('main', MainEndpoint::class)
+            ->addArgument(new LiteralArgument($this->getConfigValue('mainApiEndpoint')));
+        $container->add('auth', TokenEndpoint::class)
+            ->addArguments([
+                new LiteralArgument($this->getConfigValue('tokenApiEndpoint')),
+                'token_config',
+            ]);
+
+        // Inject the services into endpoints.
+        $container->inflector(EndpointInterface::class)
+            ->invokeMethods([
+                'setHttpClient' => [$httpClient],
```

Comment:
The inflector is a very smooth solution, however, isn't the Endpoint class too dependent on a chain of setFoo functions? Perhaps it doesn't matter that much, as an exception will be thrown when some setters are not called yet.

Reply:
it is a bit opinionated yes, an alternative would be to have the endpoint fetch what it needs and instantiation or even call each setter before usage like we do with setToken but I think this similar and more maintainable, I'd like a 3rd opinion on this